### PR TITLE
Updating base.py (parallel calibration and model #1809)

### PR DESCRIPTION
**Pull request by aashvgit**
SUMMARY:
I've added ThreadPoolExecuter To manage parallel threads and also added New Helper Method (quantize_and_update_module and updated the compress_modules making it simple and making ThreadPoolExecuter with one worker for each GPU (OPTION B)

### CODE REVIEW

**Review by aashvgit — COMMENTED** [review]
Signed-off-by: @aashvgit

File: src/llmcompressor/modifiers/quantization/gptq/base.py [hunk 1]
```diff
@@ -241,34 +242,76 @@ def compress_modules(self):
...
+                future = executor.submit(
+                    self._quantize_and_update_module,
+                    module,
+                    target_device,
+                    original_device
                 )
-                comp_logger.set_loss(loss)
+                futures.append(future)
+            for future in futures:
+                future.result()
```
[the remark this answers is not in the document]
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/quantization/gptq/base.py:272]: This might make sense, if we are allowing multi-gpu compression, modules could also be spread out across different GPUs.

**Review by brian-dellabetta — COMMENTED** [review]
Thanks for the contribution! The main maintainer for GPTQ is out of office, but I will discuss this with him next week.

**Comment by aashvgit** [comment]
Hi i would like to know if there's  any additional changes or edge cases to work with i was thinking to add multi-GPU / model-parallel test coverage if that would be helpful. should i proceed like that? @brian-dellabetta

File: src/llmcompressor/modifiers/quantization/gptq/base.py [hunk 2]
```diff
@@ -241,34 +242,76 @@ def compress_modules(self):
...
+        if not gpu_ids:
+            logger.info("No CUDA devices found. Running GPTQ sequentially on CPU.")
+            for module in modules_to_quantize:
+                self._quantize_and_update_module(module)
+            return
+
+        logger.info(f"Starting parallel GPTQ on {len(gpu_ids)} GPUs...")
+        original_device = get_execution_device(modules_to_quantize[0])
+
+        with ThreadPoolExecutor(max_workers=len(gpu_ids)) as executor:
```
Comment (brian-dellabetta) [inline src/llmcompressor/modifiers/quantization/gptq/base.py:259]: we might be able to shorten this using a different part of the concurrent.futures API. Can try this map feature instead of needing another for loop to call `future.result()`?

```python
import concurrent.futures
import time

def task_function_simple(x):
    time.sleep(1) # Simulate work
    return x * x

def main_map():
    inputs = [1, 2, 3, 4, 5]
    with concurrent.futures.ThreadPoolExecutor(max_workers=3) as executor:
        # map automatically submits tasks and yields results in order
        results = executor.map(task_function_simple, inputs)
        print("Tasks submitted. Results in order:")
        for result in results:
            print(f"Result: {result}")

if __name__ == "__main__":
    main_map()
```

**Review by brian-dellabetta — COMMENTED** [review]
Hi @aashvgit , thanks for the contribution and sorry for the slow response. I was on leave the past few weeks. Regarding your question, I left one additional comment and the gemini comment looks valid as well.

I will bring this up with the team tomorrow. We are discussing parallelization approaches and I think something like this at the level of the modifier could be a quick win for GPTQ (and most likely AWQ). Do you have any benchmarks from running this? Do you see a nice speedup?

**Comment by brian-dellabetta** [comment]
Hi @aashvgit , we spoke internally this morning about your PR. We are considering some other pathways for parallelized compression that can be applied more generically to any modifier, not specific to GPTQ, namely data-parallelism and possibly other options. It's a big consideration, and in the new year @kylesayrs will be creating an RFC for users to provide feedback. We will be sure to include this modifier-specific parallelism as one of the options, and will link in your PR. Let's leave this open and we can revisit once we have gathered some user feedback. 

We appreciate your taking the time to create this PR. I personally prefer the simplicity of your approach, but we'd have to find ways to add it to each modifier, and we need to gather more user feedback to see what path is most suitable, and we'll have to do more comprehensive benchmarking to see how much speedup can be provided (please let us know if you have any data on this). To be continued!

**Comment by aashvgit** [comment]
Hello @brian-dellabetta Thank you for the reply and also sorry for the late reply here, i think the current implementation does assumes all modules originate from the same device which i now realised that it's not a right way for multi-GPU setups maybe update the logic to track the original device per module and restore each module to respective device after that instead of using a single device. Would like you guide if the approach is correct or not maybe I'll check it and gather some data and will share the same and thanks for considering it.

**Comment by brian-dellabetta** [comment]
Hi @aashvgit , yeah some other considerations emerge when we do a lot of moving around on multiple devices, rather than stick to some of the more common flows found in training/inference like data parallelism. You can view the new RFC here:
* #2180

**Review by kylesayrs — COMMENTED** [review]
Very nice. Could you please add some performance measurements to your PR, showing the benefits of parallel computation?

FYI, we'll likely want to modify this in the future to use DP multiprocessing, but we should certainly prioritize this in the meantime!
https://github.com/vllm-project/llm-compressor/issues/2180#issuecomment-3715393706

**Comment by dsikka** [comment]
@mergifyio refresh

**Comment by HDCharles** [comment]
closing this, we have a bunch of DDP and compress_module parallelization work going on, https://github.com/vllm-project/llm-compressor/pull/2333/changes#diff-1c9652a451ec6e28d99279b8731d5e0b3c4b172e22a06aba331ad70b9674341bR354-R388 pretty much subsumes this PR
